packages/e2e-tests/specs/editor/plugins/block-context.test.js

```diff
@@ -0,0 +1,108 @@
+ */
+import { last } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import {
+	activatePlugin,
+	createNewPost,
+	deactivatePlugin,
+	insertBlock,
+	saveDraft,
+} from '@wordpress/e2e-test-utils';
+
+async function openPreviewPage( editorPage ) {
+	let openTabs = await browser.pages();
+	const expectedTabsCount = openTabs.length + 1;
+	await editorPage.click( '.editor-post-preview__button-toggle' );
+	await editorPage.waitFor( '.editor-post-preview__button-external' );
+	await editorPage.click( '.editor-post-preview__button-external' );
+
+	// Wait for the new tab to open.
+	while ( openTabs.length < expectedTabsCount ) {
+		await editorPage.waitFor( 1 );
+		openTabs = await browser.pages();
+	}
+
+	const previewPage = last( openTabs );
+	// Wait for the preview to load. We can't do interstitial detection here,
+	// because it might load too quickly for us to pick up, so we wait for
+	// the preview to load by waiting for the content to appear.
+	await previewPage.waitForSelector( '.entry-content' );
+	return previewPage;
+}
+
+describe( 'Block context', () => {
+	beforeAll( async () => {
+		await activatePlugin( 'gutenberg-test-block-context' );
+	} );
+
+	beforeEach( async () => {
+		await createNewPost();
+	} );
+
+	afterAll( async () => {
+		await deactivatePlugin( 'gutenberg-test-block-context' );
+	} );
+
+	test( 'Block context propagates to inner blocks', async () => {
+		await insertBlock( 'Test Context Provider' );
+
+		// Inserting the context provider block should select the first inner
+		// block of the template. Verify the contents of the consumer.
+		let innerBlockText = await page.evaluate(
+			() => document.activeElement.textContent
+		);
+		expect( innerBlockText ).toBe( 'The record ID is: 0' );
+
+		// Change the attribute value associated with the context.
+		await page.keyboard.press( 'ArrowUp' );
+		await page.keyboard.type( '123' );
+
+		// Verify propagated context changes.
+		await page.keyboard.press( 'ArrowDown' );
+		innerBlockText = await page.evaluate(
+			() => document.activeElement.textContent
+		);
+		expect( innerBlockText ).toBe( 'The record ID is: 123' );
+	} );
+
+	test( 'Block context is reflected in the preview', async () => {
+		await insertBlock( 'Test Context Provider' );
+		const editorPage = page;
+		const previewPage = await openPreviewPage( editorPage );
+
+		// Check default context values are populated.
+		let content = await previewPage.$eval(
+			'.entry-content',
+			( contentWrapper ) => contentWrapper.textContent.trim()
+		);
+		expect( content ).toBe( 'The record ID is: 0' );
+
+		// Return to editor to change context value to non-default.
+		await editorPage.bringToFront();
+		await editorPage.focus(
+			'[data-type="gutenberg/test-context-provider"] input'
+		);
+		await editorPage.keyboard.press( 'ArrowRight' );
+		await editorPage.keyboard.type( '123' );
+		await editorPage.waitForSelector( '.editor-post-save-draft' ); // TODO: Why is Save not available immediately?
```

Comment:
> I'd guess that React is scheduling the state update for the next frame so that the keyboard input has priority.

We have lots of other tests which call `saveDraft` immediately following some prior task 🤷 I have a feeling there might be something here with how we asynchronously render blocks in a list except in cases where the change originates from the block itself. Since the test case deals with separate InnerBlocks, it's probably prone to this.

Related: #19343

So the `waitForSelector` is probably necessary. I can update the comment with more of this context.

Reply:
Though, the "trigger" of Save becoming available should be the attribute change of the parent block, so I'm not sure why it wouldn't be synchronous 🤔

Reply:
>I'd guess that React is scheduling the state update for the next frame so that the keyboard input has priority.

This feels most plausible to me. It's not strictly React state in this case since it goes through the `setAttributes` machinery, but there's probably something in the process where React is deferring to the event. 🤷

Reply:
> I'd guess that React is scheduling the state update for the next frame so that the keyboard input has priority.

Comment updated in 8665824.